src/Mod/Sketcher/Gui/TaskSketcherTool.cpp

```diff
@@ -0,0 +1,967 @@
+        is being used, then you have to add this key to the below section such that the spinbox
+        doesn't keep the keypress event for itself. Note if you want the event to be handled by 
+        the spinbox too, you can return false.*/
+        QKeyEvent* keyEvent = static_cast<QKeyEvent*>(event);
+        if (keyEvent->key() == Qt::Key_Control)
+        {
+            sketchView->keyPressed(1, SoKeyboardEvent::LEFT_CONTROL);
+            return true;
+        }
+        else if (keyEvent->key() == Qt::Key_Shift)
+        {
+            sketchView->keyPressed(1, SoKeyboardEvent::RIGHT_SHIFT);
+            return true;
+        }
+        else if (keyEvent->key() == Qt::Key_Alt)
+        {
+            sketchView->keyPressed(1, SoKeyboardEvent::RIGHT_ALT);
+            return true;
+        }
+        else if (keyEvent->key() == Qt::Key_Escape)
+        {
+            sketchView->keyPressed(1, SoKeyboardEvent::ESCAPE);
+            return true;
+        }
+        else if (keyEvent->key() == Qt::Key_Space)
+        {
+            sketchView->keyPressed(1, SoKeyboardEvent::SPACE);
+            return true;
+        }
+    }
+    return false;
+}
+
+void SketcherToolWidget::emitTogglePreserveFilletChamferCorner(bool val) {
+    ui->checkBoxPreserveFilletChamferCorner->onSave();
+}
+void SketcherToolWidget::emitSetparameter(double val)
+{
+    isSettingSet[0] = 1;
+    toolParameters[0] = val;
+    if (unitChangeTriggeredValueChanged) {
+        unitChangeTriggeredValueChanged = 0;
+    }
+    else setParameterFocus(1);
+
+    //Make a mousemove to update geometry when parameter is set.
+    if(isWidgetActive) {
+        sketchView->mouseMove(sketchView->prvMoveCursorPos, sketchView->prvMoveViewer);
+    }
+}
+void SketcherToolWidget::emitSetparameterTwo(double val)
+{
+    isSettingSet[1] = 1;
+    toolParameters[1] = val;
+    if (unitChangeTriggeredValueChanged) {
+        unitChangeTriggeredValueChanged = 0;
+    }
+    else setParameterFocus(2);
+
+    if (isWidgetActive) {
+        sketchView->mouseMove(sketchView->prvMoveCursorPos, sketchView->prvMoveViewer);
+    }
+}
+void SketcherToolWidget::emitSetparameterThree(double val)
+{
+    isSettingSet[2] = 1;
+    toolParameters[2] = val;
+    if (unitChangeTriggeredValueChanged) {
+        unitChangeTriggeredValueChanged = 0;
+    }
+    else setParameterFocus(3);
+
+    if (isWidgetActive) {
+        sketchView->mouseMove(sketchView->prvMoveCursorPos, sketchView->prvMoveViewer);
+    }
+}
+void SketcherToolWidget::emitSetparameterFour(double val)
+{
+    isSettingSet[3] = 1;
+    toolParameters[3] = val;
+    if (unitChangeTriggeredValueChanged) {
+        unitChangeTriggeredValueChanged = 0;
+    }
+    else setParameterFocus(4);
+
+    if (isWidgetActive) {
+        sketchView->mouseMove(sketchView->prvMoveCursorPos, sketchView->prvMoveViewer);
+    }
+}
+void SketcherToolWidget::emitSetparameterFive(double val)
+{
+    isSettingSet[4] = 1;

```

Comment:
If these variables are just to keep the state, then it is a bad solution. The widget can get the values directly from the widget, there is no need to cache them and synchronise them.

Reply:
This is not to keep the state. This is to keep either the user typed a value at least once or not. See this example : 
`virtual void mouseMove(Base::Vector2d onSketchPos)
    {
        if (Mode == STATUS_SEEK_First) {
            setPositionText(onSketchPos);
            if (seekAutoConstraint(sugConstr1, onSketchPos, Base::Vector2d(0.f, 0.f))) {...}
            if (toolSettings->widget->isSettingSet[0] + toolSettings->widget->isSettingSet[1] == 2) {
                pressButton(onSketchPos);
                releaseButton(onSketchPos);
            }
        }`
You see if the user typed in the first and second parameter, then DrawSketchHandler automatically goes to STATUS_SEEK_Second by calling mouseMove which release a pressButton/releaseButton.

I'm not sure how to implement this with the new architecture.

Reply:
```
            EditCurve[0] = onSketchPos;
            if (toolSettings->widget->isSettingSet[0] == 1) {
                EditCurve[0].x = toolSettings->widget->toolParameters[0];
            }
            if (toolSettings->widget->isSettingSet[1] == 1) {
                EditCurve[0].y = toolSettings->widget->toolParameters[1];
            }
```
So I'm quite lost with the new architecture. How do the basic operation of the widget : knowing if a parameters has been typed in (isSettingSet) and getting its value (toolParameters).